### ISSUE TITLE
Issue in Astro 4.2.4+ with dynamic routes ordering

### ISSUE THREAD

**Issue by flaviocopes** [withastro/astro#10071]
### Astro Info

```block
Astro                    v4.3.5
Node                     v20.11.0
System                   macOS (arm64)
Package Manager          npm
Output                   server
Adapter                  @astrojs/node
Integrations             none
```

### Describe the Bug

In some cases, if I have a dynamic route and a non-dynamic one (think `/blog/index.astro` and `/blog/[...slug].astro`), the dynamic route takes precedence.

The `/blog` route is served by `/blog/[...slug].astro` instead of `/blog/index.astro`

NOTE: This does not happen for all folder structures. In the example I post below, the `/modules` route shows the problem.

![Screenshot-2024-02-09T03 50 15PM](https://github.com/withastro/astro/assets/11783388/2b2bbef0-2538-456d-949e-c78152729eaf)

But the `/test` route works as expected.

![Screenshot-2024-02-09T03 51 39PM](https://github.com/withastro/astro/assets/11783388/f9cd99a9-2a12-41b0-a6e8-504b83b6bf66)

And if you remove the src/pages/a folder (or any of the files in in), also /modules starts to work as expected.

Also if you rename "src/pages/test" to "src/pages/ctest" or anything that is above src/pages/modules in alphabetical order, the bug goes away.

Bug also goes away if I put a file (in alphabetical order) between those the folder `src/pages/a` and `src/pages/modules`.

[Discussed this on Discord](https://discord.com/channels/830184174198718474/1147239489861861467/1204078511640412210) where the bug was confirmed.

Works correctly up to 4.1 ✅

then tried:

```
4.2.0 ❌
4.2.1 ✅
4.2.2 ✅
4.2.3 ✅
>4.2.4 ❌
```

### What's the expected result?

The `/blog` route should be served by `/blog/index.astro` and not by `/blog/[...slug].astro` 

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-qwozj5-qxxeyv

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by Fryuni** [comment]
Thank you!

**Comment by dvartdal** [comment]
Same issue suddenly happens in my solution too, and it ruins all of my routes unfortunately...

Did you find any workaround or solution to this @flaviocopes ?

**Comment by Fryuni** [comment]
@dvartdal The issue seems to depend on other files around, some people commented that adding a dummy route in between them (name in between "index" and the name of the rest parameter) makes it work.

**Comment by flaviocopes** [comment]
thank you @Fryuni for working on this!

**Comment by dvartdal** [comment]
any updates on this @Fryuni?

**Comment by Fryuni** [comment]
The fix was merged, if it wasn't release yet it should be in the next one. Can't check right now.

**Comment by flaviocopes** [comment]
Fix released in 4.4.14
